### PR TITLE
First iteration of a data corruption feature

### PR DESCRIPTION
**Pull request by sadikneipp**
### Status
**READY**

### Todo list
- [x] Documentation
- [x] Tests added and passed
- [ ] Bump version
- [ ] Edit the changelog: ?

### Background context
First attempt to implement an artificial data corruption feature. Sort of a Chaos Monkey for datasets.

### Description of the changes proposed in the pull request
Implements a new **perturbators.py** file to store data perturbation functions and also a general function similar to `validator` in **validator.py**.

### Related PRs
None.

### Where should the reviewer start?
**Macacaos.ipynb**: simple demo of what was implemented.

**perturbator.py**: I tried to follow the same structure in validator.py: a general broad function `perturbator()` that receives specific column-wise perturbation functions like `nullify()` or `shift_mu()`.

**validator.py**: added two methods `chaos_validator` and `chaos_validator_iteration`, analogous to their non chaos counterparts but with data corruption at either train-time or test-time.

### Remaining problems or questions

Is creating the `chaos_validator` function the best approach? Maybe just an optional argument inside the original `validator`?

No logging in the new functions. How to do it?

Maybe more corruption functions.

Improve the demo. Current plot does not look that convincing on how the dataset is being degraded by nulls.

Tests.

### CODE REVIEW

File: src/fklearn/validation/perturbators.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    return col + noise
+
+@curry
+def nullify(col: pd.Series, perc: float = 1) -> pd.Series:
+    # default behavior to nullify whole column
+    n = len(col)
+    ix_to_nan = random.sample(range(n), int(n * perc))
+    ret = col.copy(deep=True)
+    ret.iloc[ix_to_nan] = np.nan
+    return ret
```
[the remark this answers is not in the document]
Comment (sadikneipp) [reply src/fklearn/validation/perturbators.py]: Done in 2c3fcbd

File: src/fklearn/validation/perturbators.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
+import numpy as np
+import random
+from toolz import curry
+import pandas as pd
+from typing import Callable, List
```
[the remark this answers is not in the document]
Comment (sadikneipp) [reply src/fklearn/validation/perturbators.py]: Addressed in 64218e2

**Comment by matheusfacure** [comment]
Nice! I like everything except Demo Notebook. Its too Nubank specific. Maybe make it a proper Demo on how to use this new feature. If don't think is worth it now you can just delete it.

File: src/fklearn/validation/validator.py [hunk 3]
```diff
@@ -104,6 +109,16 @@ def validator(train_data: pd.DataFrame,
...
 
+    perturb_fn_train : PerturbFnType
+        A partially defined corruption function that takes a dataset and returns
+        a corrupted dataset. Perturbation applied at train-time.
+    #TODO Logs?
+
+    perturb_fn_test : PerturbFnType
+        A partially defined corruption function that takes a dataset and returns
+        a corrupted dataset. Perturbation applied at test-time.
+    #TODO Logs?
```
Comment (prcastro) [inline src/fklearn/validation/validator.py]: Remove (or solve) this before merging

**Comment by matheusfacure** [comment]
Great work! But unit tests are missing

**Pull request merged**